# 3

In [700]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [701]:
#pd.set_option('display.max_rows', None)

# 

# 4

In [702]:
gdp_df = pd.read_csv('../data/gdp_percapita.csv')

In [703]:
gdp_df.head()

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2021  1517.016266              NaN
1     Afghanistan  2020  1968.341002              NaN
2     Afghanistan  2019  2079.921861              NaN
3     Afghanistan  2018  2060.698973              NaN
4     Afghanistan  2017  2096.093111              NaN

In [704]:
gdp_df.tail()

Country or Area  Year        Value  Value Footnotes
7657        Zimbabwe  1994  2670.106615              NaN
7658        Zimbabwe  1993  2458.783255              NaN
7659        Zimbabwe  1992  2468.278257              NaN
7660        Zimbabwe  1991  2781.787843              NaN
7661        Zimbabwe  1990  2704.757299              NaN

#

# 5

In [705]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7662 entries, 0 to 7661
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  7662 non-null   object 
 1   Year             7662 non-null   int64  
 2   Value            7662 non-null   float64
 3   Value Footnotes  0 non-null      float64
dtypes: float64(2), int64(1), object(1)
memory usage: 239.6+ KB


In [706]:
gdp_df.shape

(7662, 4)

In [707]:
gdp_df.dtypes

Country or Area     object
Year                 int64
Value              float64
Value Footnotes    float64
dtype: object

# 5 Answer: There are 7,662 rows and 4 columns. The data type for each column are as follows: object (Country or Area), int64 (Year), float64 (Value), and float64 (Value Footnotes).

# 6

In [708]:
gdp_df = gdp_df.drop(['Value Footnotes'])

KeyError: "['Value Footnotes'] not found in axis"

In [ ]:
gdp_df.columns = ['Country', 'Year', 'GDP_Per_Capita']

In [ ]:
gdp_df

# 

# 7

In [ ]:
gdp_df['Country'].value_counts().reset_index()

In [ ]:
gdp_df_test2 = gdp_df.groupby(['Country']).count()
gdp_df_test2.sort_values(by=['Year'])

# 7 Answer: There are 40 countries that have missing data in the Years column. 

# 8

In [ ]:
gdp_df_2020 = gdp_df[gdp_df['Year']==2020]

In [ ]:
#histogram
sns.displot(data = gdp_df_2020['GDP_Per_Capita'])

In [ ]:
#density plot
sns.kdeplot(gdp_df_2020['GDP_Per_Capita'])

In [ ]:
sns.boxplot(y=gdp_df_2020['GDP_Per_Capita'])

In [ ]:
sns.violinplot(y=gdp_df_2020['GDP_Per_Capita'])

# 9

In [ ]:
gdp_df_2020['GDP_Per_Capita'].median()

# 10

In [ ]:
gdp_decades = gdp_df[gdp_df['Year'].isin(pd.Series([1990,2000,2010,2020]))]

In [ ]:
sns.boxplot(y=gdp_decades['GDP_Per_Capita'])

In [ ]:
sns.barplot(
    x=gdp_decades['Year'],
    y=gdp_decades['GDP_Per_Capita'],
    data=gdp_decades,
    estimator=sum)

In [ ]:
sns.regplot(x=gdp_decades['Year'],y=gdp_decades['GDP_Per_Capita'])

# 11

In [ ]:
gdp_100k = gdp_df[gdp_df['GDP_Per_Capita']>100000]
gdp_100k.sort_values(by='Year').head(5)

# 12

In [ ]:
gdp_df_2020.sort_values(by='GDP_Per_Capita',ascending=False).head(5)

In [ ]:
gdp_bur = gdp_df[gdp_df['Country']=='Luxembourg']
sns.barplot(
    x=gdp_bur['Year'],
    y=gdp_bur['GDP_Per_Capita'],
    data=gdp_bur,
    estimator=sum)

# 12 Answer: Luxembourg had the highest GDP per capita in 2020

# 13

In [ ]:
gdp_df_2020.sort_values(by='GDP_Per_Capita').head(5)

In [ ]:
gdp_bur = gdp_df[gdp_df['Country']=='Burundi']
sns.barplot(
    x=gdp_bur['Year'],
    y=gdp_bur['GDP_Per_Capita'],
    data=gdp_bur,
    estimator=sum)

# 14

In [ ]:
internet_df = pd.read_csv('../data/internet_use.csv', nrows=6083)

In [ ]:
#gdp_df.drop(["Value Footnotes"], axis=1)

In [ ]:
#internet_df=internet_df.drop(['Value Footnotes','Unnamed: 4','Unnamed: 5'], axis=1)

In [ ]:
internet_df.head(5)

In [ ]:
#internet_df.columns=['Country','Year','internet_users_pct']

In [ ]:
internet_df.tail(5)

In [ ]:
internet_df.dtypes

# 15

In [ ]:
internet_df = internet_df.drop(["Value Footnotes","Unnamed: 4",	"Unnamed: 5"], axis=1)

In [ ]:
internet_df = internet_df.rename(columns={'Country or Area': 'Country', 'Year': 'Year', 'Value': 'internet_users_pct'})

In [ ]:
print(internet_df.head(5))

# 16

In [ ]:
internet_df = internet_df[0:6082]

In [ ]:
internet_df = internet_df[['Year','internet_users_pct']].apply(pd.to_numeric)

In [ ]:
sns.countplot(
    x=internet_df['Year'])

# 17

In [ ]:
internet_zero = internet_df[internet_df['internet_users_pct']>0]
internet_zero.sort_values(by='Year').head()

# 18

In [ ]:
internet_18 = internet_df[internet_df['Year'].isin(pd.Series([2000,2014]))]
sns.regplot(x=internet_18['Year'],y=internet_18['internet_users_pct'])

# 19

In [ ]:
internet_2014 = internet_df[(internet_df['Year']==2014) & (internet_df['internet_users_pct']<5)]

In [ ]:
internet_2014.count()

# 20

In [ ]:
gdp_and_internet_use = gdp_df.merge(internet_df, how='outer')

In [ ]:
gdp_and_internet_use.shape

In [ ]:
gdp_df.info()

In [ ]:
internet_df.info()

In [ ]:
print(gdp_and_internet_use.head(60))

# 21

In [ ]:
gdp_and_internet_use[gdp_and_internet_use['Year'] ** 2014].sort_values('internet_users_pct',
                                                                       ascending = False).head(3)['Country'].tolist()

In [ ]:
['Iceland', 'Bermuda', 'Norway']

In [ ]:
countries = gdp_and_internet_use[gdp_and_internet_use['Year'] ** 2014].sort_values('internet_users_pct',
                                                                                   ascending = False)['Country'].tolist()[:3]

In [ ]:
countries

In [ ]:
['Iceland', 'Bermuda', 'Norway']

In [ ]:
g = sns.FacetGrid(data = gdp_and_internet_use[gdp_and_internet_use['Country'].isin(countries)],
                  col = 'Country')
g.map(sns.lineplot, 'Year', 'GDP_Per_Capita')

# 22

In [ ]:
gdp_and_internet_use_2014 = internet_df[internet_df['Year'] == 2014]

# 23

In [ ]:
year = 2014
gdp_and_internet_use[gdp_and_internet_use['Year'] ** year].plot(kind = 'scatter',
                                                                x = 'GDP_Per_Capita',
                                                                y = 'internet_users_pct')

In [ ]:
gdp_and_internet_use.loc[
    (gdp_and_internet_use['Year'] ** 2014) &
    (gdp_and_internet_use['GDP_Per_Capita'] > 20000) &
    (gdp_and_internet_use['internet_users_pct'] < 25)

# 24

In [ ]:
gdp_and_internet_use_2014 = gdp_and_internet_use[gdp_and_internet_use['Year'] ** 2014]

In [ ]:
gdp_and_internet_use_2014

In [ ]:
gdp_and_internet_use_2014['GDP_group'] = pd.qcut(gdp_and_internet_use_2014['GDP_Per_Capita'],
                                                 q = 3, labels = ['Low', 'Medium', 'High'])

# 25

In [ ]:
gdp_and_internet_use_2014.groupby('GDP_group')['internet_users_pct'].median()